### DISCUSSION TITLE
building-a-thread-safe-random-implementation-for-dotnet-framework/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#27] [Announcements]
# building-a-thread-safe-random-implementation-for-dotnet-framework/

In this post I look at some of the ways you can misuse System.Random, comparing .NET Framework,  NET Core, and .NET 6 implementations.

https://andrewlock.net/building-a-thread-safe-random-implementation-for-dotnet-framework/?fbclid=IwAR0PC11PlAuth-sXtFr45z2DIn9b3Ig1tRgt8VefX-hQZih6NXrISdKDICI

**Comment by xilapa** [comment]
Great post, it's always good to be aware of thread safeness.

**Comment by ZOXEXIVO** [comment]
ThreadStatic is legacy way, 
ThreadLocal is more comfortable way without any locks:

```cs
public static class StaticRandom
{
    static int seed = Environment.TickCount;

    static readonly ThreadLocal<Random> random =
        new ThreadLocal<Random>(() => new Random(Interlocked.Increment(ref seed)));

    public static int Rand()
    {
        return random.Value.Next();
    }
}
```

**Reply by andrewlock** [reply]
If I remember correctly, `ThreadLocal` puts signifcantly more pressure on the GC compared to `ThreadStatic`, as it adds a finalizer. @kevingosse will put me right on that :smile:

**Comment by MarkCiliaVincenti** [comment]
Exceptional article! One thing that is not covered is what to do in case you are targeting .NET Standard 2.0. I made it all simple in a debugable micro-library at https://github.com/MarkCiliaVincenti/ThreadSafeRandomizer

You can use ThreadSafeRandom.Instance.Next and if you're on net6+ it automatically uses the Random.Shared.
